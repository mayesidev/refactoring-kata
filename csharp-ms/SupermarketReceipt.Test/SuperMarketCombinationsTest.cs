using System;
using ApprovalTests.Combinations;
using ApprovalTests.Reporters;
using SupermarketReceipt;
using Xunit;

namespace Supermarket.Test
{
    public class SupermarketCombinationsTest
    {
        [Fact]
        [UseReporter(typeof(DiffReporter))]
        public void ApproveReceiptCalculation()
        {
            var itemQuantities = new double[] {
                1,
                5
            };
            var productUnits = new[] {
                ProductUnit.Each,
                ProductUnit.Kilo
            };
            var specialOfferTypes = new[] {
                SpecialOfferType.TenPercentDiscount,
                SpecialOfferType.FiveForAmount,
                SpecialOfferType.ThreeForTwo,
                SpecialOfferType.TwoForAmount
            };
            var specialOfferArguments = new[] {
                10.0
            };

            CombinationApprovals.VerifyAllCombinations(
            CallCheckoutAndGetReceipt,
            itemQuantities,
            productUnits,
            specialOfferTypes,
            specialOfferArguments);
        }

        private string CallCheckoutAndGetReceipt(
            double itemQuantity,
            ProductUnit productUnit,
            SpecialOfferType specialOfferType,
            double specialOfferArgument)
        {
            var catalog = new FakeCatalog();
            var cart = new ShoppingCart();
            var teller = new Teller(catalog);

            var product = new Product($"chocolate", productUnit);
            catalog.AddProduct(product, 1.00);
            cart.AddItem(product);
            cart.AddItemQuantity(product, itemQuantity);
            teller.AddSpecialOffer(specialOfferType, product, specialOfferArgument);

            var receipt = teller.ChecksOutArticlesFrom(cart);
            var receiptPrinter = new ReceiptPrinter();
            receiptPrinter.StartWithNewLine();
            receiptPrinter.WrapReceiptOutputWith("-");

            var receiptOutput = receiptPrinter.PrintReceipt(receipt);
            return receiptOutput;
        }
    }
}