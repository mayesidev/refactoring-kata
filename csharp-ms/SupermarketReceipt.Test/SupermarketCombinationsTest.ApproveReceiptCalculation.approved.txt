[1,Each,TenPercentDiscount,10] => 
----------------------------------------
chocolate                           1.00
chocolate                           1.00
10% off(chocolate)                 -0.20

Total:                              1.80
----------------------------------------

[1,Each,FiveForAmount,10] => 
----------------------------------------
chocolate                           1.00
chocolate                           1.00

Total:                              2.00
----------------------------------------

[1,Each,ThreeForTwo,10] => 
----------------------------------------
chocolate                           1.00
chocolate                           1.00

Total:                              2.00
----------------------------------------

[1,Each,TwoForAmount,10] => 
----------------------------------------
chocolate                           1.00
chocolate                           1.00
2 for 10(chocolate)                 8.00

Total:                             10.00
----------------------------------------

[1,Kilo,TenPercentDiscount,10] => 
----------------------------------------
chocolate                           1.00
chocolate                           1.00
10% off(chocolate)                 -0.20

Total:                              1.80
----------------------------------------

[1,Kilo,FiveForAmount,10] => 
----------------------------------------
chocolate                           1.00
chocolate                           1.00

Total:                              2.00
----------------------------------------

[1,Kilo,ThreeForTwo,10] => 
----------------------------------------
chocolate                           1.00
chocolate                           1.00

Total:                              2.00
----------------------------------------

[1,Kilo,TwoForAmount,10] => 
----------------------------------------
chocolate                           1.00
chocolate                           1.00
2 for 10(chocolate)                 8.00

Total:                             10.00
----------------------------------------

[5,Each,TenPercentDiscount,10] => 
----------------------------------------
chocolate                           1.00
chocolate                           5.00
  1.00 * 5
10% off(chocolate)                 -0.60

Total:                              5.40
----------------------------------------

[5,Each,FiveForAmount,10] => 
----------------------------------------
chocolate                           1.00
chocolate                           5.00
  1.00 * 5
5 for 10(chocolate)                 5.00

Total:                             11.00
----------------------------------------

[5,Each,ThreeForTwo,10] => 
----------------------------------------
chocolate                           1.00
chocolate                           5.00
  1.00 * 5
3 for 2(chocolate)                 -2.00

Total:                              4.00
----------------------------------------

[5,Each,TwoForAmount,10] => 
----------------------------------------
chocolate                           1.00
chocolate                           5.00
  1.00 * 5
2 for 10(chocolate)                24.00

Total:                             30.00
----------------------------------------

[5,Kilo,TenPercentDiscount,10] => 
----------------------------------------
chocolate                           1.00
chocolate                           5.00
  1.00 * 5.000
10% off(chocolate)                 -0.60

Total:                              5.40
----------------------------------------

[5,Kilo,FiveForAmount,10] => 
----------------------------------------
chocolate                           1.00
chocolate                           5.00
  1.00 * 5.000
5 for 10(chocolate)                 5.00

Total:                             11.00
----------------------------------------

[5,Kilo,ThreeForTwo,10] => 
----------------------------------------
chocolate                           1.00
chocolate                           5.00
  1.00 * 5.000
3 for 2(chocolate)                 -2.00

Total:                              4.00
----------------------------------------

[5,Kilo,TwoForAmount,10] => 
----------------------------------------
chocolate                           1.00
chocolate                           5.00
  1.00 * 5.000
2 for 10(chocolate)                24.00

Total:                             30.00
----------------------------------------

